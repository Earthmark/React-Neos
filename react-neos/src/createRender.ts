import createReconciler, { OutboundSignal } from "./renderer";
import { elementDefs } from "./components";
import { InboundSignal } from "./signalFormatter";

export interface ReactNeosRenderHandler {
  bind(): {
    render(signal?: Array<InboundSignal>): Array<OutboundSignal>;
  };
}

export default function createRender(
  node: React.ReactNode
): ReactNeosRenderHandler {
  const reconciler = createReconciler(elementDefs);
  return {
    bind: () => {
      const b = reconciler.bind();
      return {
        render: (signal) => b.render(node, signal),
      };
    },
  };
}
