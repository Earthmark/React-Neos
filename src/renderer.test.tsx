import React from 'react';
import createRender from "./renderer";
import n from "./components";

interface Controller {
  toggleCanvas?: boolean,
  objects: Array<{
    id: string
  }>
}

const TestComponent = ({routeController}:
  {routeController: (provider: (newController: Controller) => void) => void}) => {

  const [c, setC] = React.useState<Controller>(() => ({objects:[]}));

  routeController(setC);

  return <n.transform position={{x: 2, y: 4, z: 19}}>
    {
      c.toggleCanvas ? <n.smoothTransform>
        <n.box />
      </n.smoothTransform> : <n.canvas>
        <n.rect>
          <n.text>
            This contains text!
          </n.text>
        </n.rect>
      </n.canvas>
    }
    <n.canvas>
      {
        c.objects.map(o => 
          <n.text key={o.id}>
            {o.id}
          </n.text>)
      }
    </n.canvas>
  </n.transform>
}

test("Verify hierarchy shows as expected", () => {
  var updater:  (newController: Controller) => void = ()=>{};

  const b = {
    toggleCanvas: false,
    objects: [{id: "1"}]
  };
  const renderer = createRender(<TestComponent routeController={c => updater = c }/>);
  const instance = renderer.createInstance();

  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  updater(b);
  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  updater({...b, objects: [{id: "1"}, {id: "2"}]});
  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  updater({...b, objects: [{id: "2"}]});
  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  updater({...b, objects: [{id: "1"}, {id: "2"}]});
  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  updater({...b, toggleCanvas: true});
  expect(instance.render()).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);

  expect(instance.render([
    {
      signal: "event",
      id: "1",
      event: "click",
      arg: "payload"
    }
  ])).toMatchSnapshot();
  expect(instance.render()).toStrictEqual([]);
});

test("unexpected components raise errors", () => {
  const renderer = createRender(<div/>);
  const instance = renderer.createInstance();
  expect(() => instance.render()).toThrowError();
});

test("components raise errors when they unexpectedly contain text", () => {
  const renderer = createRender(<n.transform>{"I'm illegal!" as any}</n.transform>);
  const instance = renderer.createInstance();
  expect(() => instance.render()).toThrowError();
});