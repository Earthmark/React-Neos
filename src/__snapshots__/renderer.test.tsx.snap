// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refs Interconnect 1`] = `
Array [
  Object {
    "id": "1",
    "signal": "create",
    "type": "unlitMaterial",
  },
  Object {
    "id": "1",
    "props": Array [
      Object {
        "prop": "color",
        "type": "color",
        "value": "[1;0;1;1]",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "2",
    "signal": "create",
    "type": "renderer",
  },
  Object {
    "id": "3",
    "signal": "create",
    "type": "transform",
  },
  Object {
    "id": "3",
    "props": Array [
      Object {
        "prop": "position",
        "type": "float3",
        "value": "[2;4;19]",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "2",
    "parentId": "3",
    "signal": "setParent",
  },
  Object {
    "id": "root",
    "signal": "remove",
  },
  Object {
    "id": "1",
    "parentId": "root",
    "signal": "setParent",
  },
  Object {
    "id": "3",
    "parentId": "root",
    "signal": "setParent",
  },
  Object {
    "id": "2",
    "props": Array [
      Object {
        "prop": "material",
        "type": "ref",
        "value": "1.self",
      },
    ],
    "signal": "update",
  },
]
`;

exports[`Verify hierarchy shows as expected 1`] = `
Array [
  Object {
    "id": "1",
    "signal": "create",
    "type": "text",
  },
  Object {
    "id": "1",
    "props": Array [
      Object {
        "prop": "children",
        "type": "string",
        "value": "This%20contains%20text!",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "2",
    "signal": "create",
    "type": "text",
  },
  Object {
    "id": "2",
    "props": Array [
      Object {
        "prop": "children",
        "type": "string",
        "value": "1",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "3",
    "signal": "create",
    "type": "canvas",
  },
  Object {
    "id": "2",
    "parentId": "3",
    "signal": "setParent",
  },
  Object {
    "id": "4",
    "signal": "create",
    "type": "transform",
  },
  Object {
    "id": "4",
    "props": Array [
      Object {
        "prop": "position",
        "type": "float3",
        "value": "[2;4;19]",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "1",
    "parentId": "4",
    "signal": "setParent",
  },
  Object {
    "id": "3",
    "parentId": "4",
    "signal": "setParent",
  },
  Object {
    "id": "5",
    "signal": "create",
    "type": "transform",
  },
  Object {
    "id": "root",
    "signal": "remove",
  },
  Object {
    "id": "4",
    "parentId": "root",
    "signal": "setParent",
  },
  Object {
    "id": "5",
    "parentId": "root",
    "signal": "setParent",
  },
]
`;

exports[`Verify hierarchy shows as expected 2`] = `
Array [
  Object {
    "id": "6",
    "signal": "create",
    "type": "box",
  },
  Object {
    "after": "5",
    "id": "6",
    "parentId": "root",
    "signal": "setParent",
  },
]
`;

exports[`Verify hierarchy shows as expected 3`] = `
Array [
  Object {
    "id": "6",
    "signal": "remove",
  },
]
`;

exports[`Verify hierarchy shows as expected 4`] = `
Array [
  Object {
    "id": "2",
    "props": Array [
      Object {
        "prop": "anchorMin",
        "type": "float2",
        "value": "[2;4]",
      },
    ],
    "signal": "update",
  },
]
`;

exports[`Verify hierarchy shows as expected 5`] = `
Array [
  Object {
    "id": "2",
    "props": Array [
      Object {
        "prop": "anchorMin",
        "type": "float2",
        "value": null,
      },
    ],
    "signal": "update",
  },
]
`;

exports[`Verify hierarchy shows as expected 6`] = `
Array [
  Object {
    "id": "7",
    "signal": "create",
    "type": "text",
  },
  Object {
    "id": "7",
    "props": Array [
      Object {
        "prop": "children",
        "type": "string",
        "value": "2",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "8",
    "signal": "create",
    "type": "text",
  },
  Object {
    "id": "8",
    "props": Array [
      Object {
        "prop": "children",
        "type": "string",
        "value": "3",
      },
    ],
    "signal": "update",
  },
  Object {
    "id": "7",
    "parentId": "3",
    "signal": "setParent",
  },
  Object {
    "id": "8",
    "parentId": "3",
    "signal": "setParent",
  },
]
`;

exports[`Verify hierarchy shows as expected 7`] = `
Array [
  Object {
    "id": "7",
    "signal": "remove",
  },
]
`;

exports[`Verify hierarchy shows as expected 8`] = `
Array [
  Object {
    "id": "9",
    "signal": "create",
    "type": "text",
  },
  Object {
    "id": "9",
    "props": Array [
      Object {
        "prop": "children",
        "type": "string",
        "value": "2",
      },
    ],
    "signal": "update",
  },
  Object {
    "after": "8",
    "id": "9",
    "parentId": "3",
    "signal": "setParent",
  },
]
`;

exports[`Verify hierarchy shows as expected 9`] = `
Array [
  Object {
    "id": "10",
    "signal": "create",
    "type": "box",
  },
  Object {
    "id": "1",
    "signal": "remove",
  },
  Object {
    "after": "3",
    "id": "10",
    "parentId": "4",
    "signal": "setParent",
  },
]
`;

exports[`Verify hierarchy shows as expected 10`] = `Array []`;
